backend/http/http.go

```diff
@@ -0,0 +1,345 @@
+	if loader == nil {
+		loader = DefaultLoader
+	}
+	urlPath := r.URL.Path
+	urlPath = strings.TrimPrefix(urlPath, h.Prefix)
+	for _, s := range services {
+		if m := s.pattern.FindStringSubmatch(urlPath); m != nil {
+			if r.Method != s.method {
+				renderStatusError(w, http.StatusMethodNotAllowed)
+				return
+			}
+
+			repo := strings.TrimPrefix(m[1], "/")
+			file := strings.Replace(urlPath, repo+"/", "", 1)
+			ep, err := transport.NewEndpoint(repo)
+			if err != nil {
+				logf(h.ErrorLog, "error creating endpoint: %v", err)
+				renderStatusError(w, http.StatusBadRequest)
+				return
+			}
+
+			st, err := loader.Load(ep)
+			if err != nil {
+				logf(h.ErrorLog, "error loading repository: %v", err)
+				renderStatusError(w, http.StatusNotFound)
+				return
+			}
+
+			ctx := r.Context()
+			ctx = context.WithValue(ctx, contextKey("errorLog"), h.ErrorLog)
+			ctx = context.WithValue(ctx, contextKey("repo"), m[1])
+			ctx = context.WithValue(ctx, contextKey("file"), file)
+			ctx = context.WithValue(ctx, contextKey("service"), s.svc)
+			ctx = context.WithValue(ctx, contextKey("storer"), st)
+			ctx = context.WithValue(ctx, contextKey("endpoint"), ep)
+
+			s.handler(w, r.WithContext(ctx))
+			return
+		}
+	}
+
+	// If no service matched, return 404.
+	renderStatusError(w, http.StatusNotFound)
+}
+
+
+func serviceRpc(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	st, ok := ctx.Value(contextKey("storer")).(storage.Storer)
+	if !ok {
+		renderStatusError(w, http.StatusInternalServerError)
+		return
+	}
+	svc, ok := ctx.Value(contextKey("service")).(transport.Service)
+	if !ok {
+		renderStatusError(w, http.StatusInternalServerError)
+		return
+	}
+	errorLog, ok := ctx.Value(contextKey("errorLog")).(*log.Logger)
+	if !ok {
+		renderStatusError(w, http.StatusInternalServerError)
+		return
+	}
+	version := r.Header.Get("Git-Protocol")
+	contentType := strings.ToLower(strings.TrimSpace(r.Header.Get("Content-Type")))
+
+	expectedContentType := fmt.Sprintf("application/x-git-%s-request", svc.Name())
+	if contentType != expectedContentType {
+		renderStatusError(w, http.StatusForbidden)
+		return
+	}
+
+	w.Header().Set("Content-Type", fmt.Sprintf("application/x-git-%s-result", svc.Name()))
+	w.Header().Set("Connection", "Keep-Alive")
+	w.Header().Set("Transfer-Encoding", "chunked")
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+
+	var reader io.ReadCloser
+	var err error
+	switch r.Header.Get("Content-Encoding") {
+	case "gzip":
+		reader, err = gzip.NewReader(r.Body)
+		if err != nil {
+			logf(errorLog, "error creating gzip reader: %v", err)
+			renderStatusError(w, http.StatusInternalServerError)
+			return
+		}
+		defer reader.Close() //nolint:errcheck
+	default:
+		reader = r.Body
+	}
+
+	frw := &flushResponseWriter{ResponseWriter: w, log: errorLog}
+
+	switch svc {
+	case transport.UploadPackService:
+		err = transport.UploadPack(ctx, st, reader, frw,
+			&transport.UploadPackOptions{
+				GitProtocol:   version,
```

Comment:
Should we verify the version before passing it? I know it's verified inside, but what if something changes inside?

Reply:
I don't think that's necessary, as this middleware is just a frontend to the actual server implementation. If something changes in `UploadPack`/`ReceivePack`, we'd also need to update the logic everywhere these routines are used.